src/libsystemd-network/dhcp6-option.c

```diff
@@ -370,47 +370,42 @@ int dhcp6_option_append_vendor_class(uint8_t **buf, size_t *buflen, char * const
         return dhcp6_option_append(buf, buflen, SD_DHCP6_OPTION_VENDOR_CLASS, total, p);
 }
 
-static int option_parse_hdr(uint8_t **buf, size_t *buflen, uint16_t *optcode, size_t *optlen) {
-        DHCP6Option *option = (DHCP6Option*) *buf;
-        uint16_t len;
-
-        assert_return(buf, -EINVAL);
-        assert_return(optcode, -EINVAL);
-        assert_return(optlen, -EINVAL);
-
-        if (*buflen < offsetof(DHCP6Option, data))
-                return -ENOMSG;
+int dhcp6_option_parse(
+                const uint8_t *buf,
+                size_t buflen,
+                size_t *offset,
+                uint16_t *ret_option_code,
+                size_t *ret_option_data_len,
+                const uint8_t **ret_option_data) {
 
-        len = be16toh(option->len);
+        const DHCP6Option *option;
+        size_t len;
 
-        if (len > *buflen)
-                return -ENOMSG;
+        assert(buf);
+        assert(offset);
+        assert(ret_option_code);
+        assert(ret_option_data_len);
+        assert(ret_option_data);
 
-        *optcode = be16toh(option->code);
-        *optlen = len;
+        if (buflen < offsetof(DHCP6Option, data))
+                return -EBADMSG;
 
-        *buf += 4;
-        *buflen -= 4;
+        if (*offset >= buflen)
+                return -EBADMSG;
 
-        return 0;
-}
+        if (*offset + offsetof(DHCP6Option, data) >= buflen)
```

Comment:
slightly safer would be:

```c
if (*offset >= buflen - offsetof(DHCPOption, data))
```

why that? because the addition might cause overflow theoretically, but the subtraction cannot, since you already made sure of that via the if check further up.